docs/tutorials/basic/data.md

```diff
@@ -122,17 +126,17 @@ class SimpleIter(mx.io.DataIter):
 
     @property
     def provide_data(self):
-        return self._provide_data
+        return zip(self._data_names, self._data_shapes)
 
     @property
     def provide_label(self):
-        return self._provide_label
+        return zip(self._label_names, self._label_shapes)
```

Comment:
?

Reply:
Can you be more specific about your comment here ? :-)

The zip function in python2 used to provide a list of tuples which are not exhausted if you have traversed through them once. In python3 zip returns an iterator which gets "exhausted" if we have traversed through once. We, therefore, are trying to create a new iterator (via zip) for every batch.

Reply:
Isn't this already done in `__init__`? 
`self._provide_data = zip(data_names, data_shapes)`

Reply:
Is there a guarantee that the functions _provide_data_ and _provide_label_ will be called only once by the caller? If yes, then we don't need to make the change to the function. If no, then we'll have to explicitly call zip every time.

In [Python2, zip](https://docs.python.org/2/library/functions.html#zip) returns a list. So, the second time you call _provide\_data_ fn, the list is used again.
However, in [Python 3, zip](https://docs.python.org/3/library/functions.html#zip) it returns an iterator. The second time you call _provide\_data_ fn, the iterator throws an error since we already exhausted the iterator the first time we called _provide\_data_ fn. We, therefore, need to re-initialize via zip everytime we call _provide\_data_.